File: Go/trie_utils/trie_utils_test.go
package trie_utils

import (
	"sort"
	"testing"
)

// ============================================================================
// Trie Creation Tests
// ============================================================================

func TestNewTrie(t *testing.T) {
	trie := NewTrie()
	if trie == nil {
		t.Fatal("NewTrie returned nil")
	}
	if trie.Size() != 0 {
		t.Errorf("New trie should be empty, got size %d", trie.Size())
	}
	if !trie.IsEmpty() {
		t.Error("New trie should report as empty")
	}
	if !trie.caseSensitive {
		t.Error("NewTrie should be case-sensitive by default")
	}
}

func TestNewTrieCaseInsensitive(t *testing.T) {
	trie := NewTrieCaseInsensitive()
	if trie == nil {
		t.Fatal("NewTrieCaseInsensitive returned nil")
	}
	if trie.caseSensitive {
		t.Error("NewTrieCaseInsensitive should be case-insensitive")
	}
}

// ============================================================================
// Insert Tests
// ============================================================================

func TestInsert(t *testing.T) {
	trie := NewTrie()

	// Insert new word
	if !trie.Insert("hello") {
		t.Error("Insert of new word should return true")
	}
	if trie.Size() != 1 {
		t.Errorf("Expected size 1, got %d", trie.Size())
	}

	// Insert same word again
	if trie.Insert("hello") {
		t.Error("Insert of existing word should return false")
	}
	if trie.Size() != 1 {
		t.Errorf("Size should still be 1, got %d", trie.Size())
	}

	// Insert empty string
	if trie.Insert("") {
		t.Error("Insert of empty string should return false")
	}
}

func TestInsertCaseSensitive(t *testing.T) {
	trie := NewTrie()

	trie.Insert("hello")
	trie.Insert("Hello")
	trie.Insert("HELLO")

	if trie.Size() != 3 {
		t.Errorf("Expected 3 distinct words (case-sensitive), got %d", trie.Size())
	}

	if !trie.Search("hello") || !trie.Search("Hello") || !trie.Search("HELLO") {
		t.Error("All case variations should exist")
	}
}

func TestInsertCaseInsensitive(t *testing.T) {
	trie := NewTrieCaseInsensitive()

	trie.Insert("hello")
	trie.Insert("Hello")
	trie.Insert("HELLO")

	if trie.Size() != 1 {
		t.Errorf("Expected 1 word (case-insensitive), got %d", trie.Size())
	}

	if !trie.Search("hello") || !trie.Search("Hello") || !trie.Search("HELLO") {
		t.Error("All case variations should match")
	}
}

func TestInsertBatch(t *testing.T) {
	trie := NewTrie()
	words := []string{"apple", "app", "application", "apply", "banana"}

	inserted := trie.InsertBatch(words)
	if inserted != 5 {
		t.Errorf("Expected 5 insertions, got %d", inserted)
	}

	// Insert duplicates
	inserted = trie.InsertBatch(words)
	if inserted != 0 {
		t.Errorf("Expected 0 new insertions, got %d", inserted)
	}
}

// ============================================================================
// Search Tests
// ============================================================================

func TestSearch(t *testing.T) {
	trie := NewTrie()
	words := []string{"apple", "app", "application", "banana"}

	for _, w := range words {
		trie.Insert(w)
	}

	for _, w := range words {
		if !trie.Search(w) {
			t.Errorf("Should find word: %s", w)
		}
	}

	// Non-existent words
	if trie.Search("ap") {
		t.Error("'ap' should not exist as a word")
	}
	if trie.Search("applex") {
		t.Error("'applex' should not exist")
	}
	if trie.Search("") {
		t.Error("Empty string should not exist")
	}
}

func TestGetCount(t *testing.T) {
	trie := NewTrie()

	trie.Insert("hello")
	trie.Insert("hello")
	trie.Insert("hello")

	if count := trie.GetCount("hello"); count != 3 {
		t.Errorf("Expected count 3, got %d", count)
	}

	if count := trie.GetCount("world"); count != 0 {
		t.Errorf("Expected count 0 for non-existent word, got %d", count)
	}
}

// ============================================================================
// Prefix Tests
// ============================================================================

func TestStartsWith(t *testing.T) {
	trie := NewTrie()
	trie.InsertBatch([]string{"apple", "app", "application", "banana"})

	tests := []struct {
		prefix   string
		expected bool
	}{
		{"app", true},
		{"apple", true},
		{"ban", true},
		{"orange", false},
		{"x", false},
		{"", true}, // Empty prefix matches all
	}

	for _, tt := range tests {
		t.Run(tt.prefix, func(t *testing.T) {
			if result := trie.StartsWith(tt.prefix); result != tt.expected {
				t.Errorf("StartsWith(%q) = %v, want %v", tt.prefix, result, tt.expected)
			}
		})
	}
}

func TestLongestCommonPrefix(t *testing.T) {
	tests := []struct {
		name     string
		words    []string
		expected string
	}{
		{"common prefix", []string{"apple", "app", "application"}, "app"},
		{"no common prefix", []string{"cat", "dog", "bird"}, ""},
		{"single word", []string{"hello"}, "hello"},
		{"empty trie", []string{}, ""},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			trie := NewTrie()
			trie.InsertBatch(tt.words)

			if result := trie.LongestCommonPrefix(); result != tt.expected {
				t.Errorf("LongestCommonPrefix() = %q, want %q", result, tt.expected)
			}
		})
	}
}

func TestLongestPrefixOf(t *testing.T) {
	trie := NewTrie()
	trie.InsertBatch([]string{"app", "apple", "application"})

	tests := []struct {
		word     string
		expected string
	}{
		{"applepie", "apple"},
		{"application", "application"},
		{"appetite", "app"},
		{"banana", ""},
		{"", ""},
	}

	for _, tt := range tests {
		t.Run(tt.word, func(t *testing.T) {
			if result := trie.LongestPrefixOf(tt.word); result != tt.expected {
				t.Errorf("LongestPrefixOf(%q) = %q, want %q", tt.word, result, tt.expected)
			}
		})
	}
}

func TestCountPrefix(t *testing.T) {
	trie := NewTrie()
	trie.InsertBatch([]string{"apple", "app", "application", "apply", "banana"})

	tests := []struct {
		prefix   string
		expected int
	}{
		{"app", 4},
		{"apple", 1},
		{"ban", 1},
		{"xyz", 0},
		{"", 5},
	}

	for _, tt := range tests {
		t.Run(tt.prefix, func(t *testing.T) {
			if result := trie.CountPrefix(tt.prefix); result != tt.expected {
				t.Errorf("CountPrefix(%q) = %d, want %d", tt.prefix, result, tt.expected)
			}
		})
	}
}

// ============================================================================
// Delete Tests
// ============================================================================

func TestDelete(t *testing.T) {
	trie := NewTrie()
	trie.InsertBatch([]string{"apple", "app", "application"})

	// Delete existing word
	if !trie.Delete("app") {
		t.Error("Delete of existing word should return true")
	}
	if trie.Search("app") {
		t.Error("'app' should not exist after deletion")
	}
	if trie.Size() != 2 {
		t.Errorf("Size should be 2, got %d", trie.Size())
	}

	// Other words should still exist
	if !trie.Search("apple") || !trie.Search("application") {
		t.Error("'apple' and 'application' should still exist")
	}

	// Delete non-existent word
	if trie.Delete("xyz") {
		t.Error("Delete of non-existent word should return false")
	}

	// Delete empty string
	if trie.Delete("") {
		t.Error("Delete of empty string should return false")
	}
}

func TestDeleteCleanup(t *testing.T) {
	trie := NewTrie()
	trie.Insert("apple")
	trie.Insert("app")

	// Delete 'apple' should not delete 'app'
	trie.Delete("apple")
	if !trie.Search("app") {
		t.Error("'app' should still exist after deleting 'apple'")
	}
	if trie.StartsWith("appl") {
		t.Error("'appl' prefix should not exist after deleting 'apple'")
	}
}

// ============================================================================
// AutoComplete Tests
// ============================================================================

func TestAutoComplete(t *testing.T) {
	trie := NewTrie()
	trie.InsertBatch([]string{"apple", "app", "application", "apply", "banana", "band"})

	tests := []struct {
		name     string
		prefix   string
		expected []string
	}{
		{"app prefix", "app", []string{"app", "apple", "application", "apply"}},
		{"ban prefix", "ban", []string{"banana", "band"}},
		{"no matches", "xyz", []string{}},
		{"empty prefix", "", []string{"app", "apple", "application", "apply", "banana", "band"}},
		{"exact match", "apple", []string{"apple"}},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := trie.AutoComplete(tt.prefix)
			sort.Strings(tt.expected)
			if !equalSlices(result, tt.expected) {
				t.Errorf("AutoComplete(%q) = %v, want %v", tt.prefix, result, tt.expected)
			}
		})
	}
}

func TestAutoCompleteN(t *testing.T) {
	trie := NewTrie()
	trie.InsertBatch([]string{"apple", "app", "application", "apply"})

	result := trie.AutoCompleteN("app", 2)
	if len(result) != 2 {
		t.Errorf("Expected 2 results, got %d", len(result))
	}
}

func TestAutoCompleteByFrequency(t *testing.T) {
	trie := NewTrie()
	
	// Insert with different frequencies
	trie.Insert("apple")
	trie.Insert("apple")
	trie.Insert("apple")
	trie.Insert("app")
	trie.Insert("app")
	trie.Insert("application")

	result := trie.AutoCompleteByFrequency("app", 10)
	
	// 'apple' should be first (count=3), then 'app' (count=2), then 'application' (count=1)
	if len(result) < 3 {
		t.Fatalf("Expected at least 3 results, got %d", len(result))
	}
	if result[0] != "apple" {
		t.Errorf("Expected 'apple' first, got %q", result[0])
	}
	if result[1] != "app" {
		t.Errorf("Expected 'app' second, got %q", result[1])
	}
}

// ============================================================================
// Pattern Matching Tests
// ============================================================================

func TestMatchPattern(t *testing.T) {
	trie := NewTrie()
	trie.InsertBatch([]string{"cat", "bat", "rat", "car", "bar", "cart", "cats"})

	tests := []struct {
		name     string
		pattern  string
		expected []string
	}{
		{"single wildcard", "?at", []string{"bat", "cat", "rat"}},
		{"prefix wildcard", "c??", []string{"car", "cat"}},
		{"suffix wildcard", "ca?", []string{"car", "cat"}},
		{"star prefix", "*", []string{"bar", "bat", "car", "cart", "cat", "cats", "rat"}},
		{"star middle", "c*t", []string{"cat", "cart", "cats"}},
		{"no match", "z??", []string{}},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := trie.MatchPattern(tt.pattern)
			sort.Strings(tt.expected)
			if !equalSlices(result, tt.expected) {
				t.Errorf("MatchPattern(%q) = %v, want %v", tt.pattern, result, tt.expected)
			}
		})
	}
}

// ============================================================================
// Fuzzy Search Tests
// ============================================================================

func TestFuzzySearch(t *testing.T) {
	trie := NewTrie()
	trie.InsertBatch([]string{"hello", "help", "held", "helmet", "world", "word"})

	tests := []struct {
		name     string
		word     string
		distance int
	}{
		{"exact match", "hello", 0},
		{"one edit", "helio", 1},
		{"two edits", "hallo", 2},
		{"three edits", "hella", 1},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			results := trie.FuzzySearch(tt.word, tt.distance)
			// Verify all results are within the specified distance
			for _, r := range results {
				d := levenshteinDistance(tt.word, r)
				if d > tt.distance {
					t.Errorf("FuzzySearch returned %q with distance %d, max allowed %d", r, d, tt.distance)
				}
			}
		})
	}
}

// ============================================================================
// Statistics Tests
// ============================================================================

func TestGetStats(t *testing.T) {
	trie := NewTrie()
	trie.InsertBatch([]string{"a", "ab", "abc", "abcd"})

	stats := trie.GetStats()

	if stats.TotalWords != 4 {
		t.Errorf("Expected 4 words, got %d", stats.TotalWords)
	}
	if stats.TotalNodes < 4 {
		t.Errorf("Expected at least 4 nodes, got %d", stats.TotalNodes)
	}
	if stats.MaxDepth != 4 {
		t.Errorf("Expected max depth 4, got %d", stats.MaxDepth)
	}
}

func TestGetStatsEmpty(t *testing.T) {
	trie := NewTrie()
	stats := trie.GetStats()

	if stats.TotalWords != 0 {
		t.Errorf("Expected 0 words, got %d", stats.TotalWords)
	}
}

// ============================================================================
// Clear Tests
// ============================================================================

func TestClear(t *testing.T) {
	trie := NewTrie()
	trie.InsertBatch([]string{"apple", "banana", "cherry"})

	trie.Clear()

	if trie.Size() != 0 {
		t.Errorf("Size should be 0 after clear, got %d", trie.Size())
	}
	if !trie.IsEmpty() {
		t.Error("Trie should be empty after clear")
	}
	if trie.Search("apple") {
		t.Error("Should not find words after clear")
	}
}

// ============================================================================
// GetAllWords Tests
// ============================================================================

func TestGetAllWords(t *testing.T) {
	trie := NewTrie()
	words := []string{"delta", "alpha", "charlie", "bravo"}
	trie.InsertBatch(words)

	result := trie.GetAllWords()
	expected := []string{"alpha", "bravo", "charlie", "delta"}

	if !equalSlices(result, expected) {
		t.Errorf("GetAllWords() = %v, want %v", result, expected)
	}
}

// ============================================================================
// Helper Functions
// ============================================================================

func equalSlices(a, b []string) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

// ============================================================================
// Benchmarks
// ============================================================================

func BenchmarkInsert(b *testing.B) {
	trie := NewTrie()
	for i := 0; i < b.N; i++ {
		trie.Insert("benchmark_word")
		trie.Clear()
	}
}

func BenchmarkSearch(b *testing.B) {
	trie := NewTrie()
	trie.Insert("benchmark_word")
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		trie.Search("benchmark_word")
	}
}

func BenchmarkAutoComplete(b *testing.B) {
	trie := NewTrie()
	words := []string{"apple", "app", "application", "apply", "aptitude", "append"}
	trie.InsertBatch(words)
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		trie.AutoComplete("app")
	}
}

func BenchmarkFuzzySearch(b *testing.B) {
	trie := NewTrie()
	// Insert a larger set of words
	for i := 0; i < 1000; i++ {
		trie.Insert(generateWord(i))
	}
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		trie.FuzzySearch("test", 2)
	}
}

func generateWord(n int) string {
	const letters = "abcdefghijklmnopqrstuvwxyz"
	word := ""
	for i := 0; i < 5; i++ {
		word += string(letters[(n+i)%26])
	}
	return word
}

File: C#/trie_utils/TrieUtilsTest.cs
using System;
using System.Collections.Generic;
using System.Linq;
using AllToolkit.TrieUtils;

namespace AllToolkit.Tests
{
    /// <summary>
    /// TrieUtils 单元测试
    /// </summary>
    class TrieUtilsTest
    {
        private static int passed = 0;
        private static int failed = 0;

        static void Main(string[] args)
        {
            Console.WriteLine("Running TrieUtils tests...\n");

            TestBasicOperations();
            TestSearchOperations();
            TestPrefixOperations();
            TestDeleteOperations();
            TestAutoComplete();
            TestAdvancedFeatures();
            TestGenericTrie();
            TestUnicodeSupport();
            TestConcurrency();
            TestEdgeCases();

            Console.WriteLine("\n========================================");
            Console.WriteLine($"Results: {passed} passed, {failed} failed");
            Console.WriteLine("========================================");

            Environment.Exit(failed > 0 ? 1 : 0);
        }

        static void Assert(bool condition, string testName)
        {
            if (condition)
            {
                Console.WriteLine($"  ✓ {testName}");
                passed++;
            }
            else
            {
                Console.WriteLine($"  ✗ {testName}");
                failed++;
            }
        }

        static void TestBasicOperations()
        {
            Console.WriteLine("\nBasic Operations Tests:");

            using (var trie = new Trie())
            {
                Assert(trie.Size == 0, "New trie has size 0");
                Assert(trie.IsEmpty, "New trie is empty");

                Assert(trie.Insert("hello"), "Insert new word returns true");
                Assert(trie.Size == 1, "Size is 1 after insert");

                Assert(!trie.Insert("hello"), "Insert duplicate returns false");
                Assert(trie.Size == 1, "Size unchanged after duplicate");

                Assert(trie.Insert("help"), "Insert another word");
                Assert(trie.Size == 2, "Size is 2");

                Assert(!trie.Insert(""), "Insert empty string returns false");
                Assert(!trie.Insert(null), "Insert null returns false");
            }
        }

        static void TestSearchOperations()
        {
            Console.WriteLine("\nSearch Operations Tests:");

            using (var trie = new Trie())
            {
                trie.Insert("hello");
                trie.Insert("help");
                trie.Insert("world");

                Assert(trie.Search("hello"), "Search existing word");
                Assert(trie.Search("help"), "Search another existing word");
                Assert(!trie.Search("hel"), "Search prefix not word");
                Assert(!trie.Search("helloworld"), "Search non-existing word");
                Assert(!trie.Search(""), "Search empty string");

                // Search with value
                trie.Insert("apple", 1);
                trie.Insert("banana", "fruit");

                object value;
                Assert(trie.SearchWithValue("apple", out value) && (int)value == 1, "SearchWithValue int");
                Assert(trie.SearchWithValue("banana", out value) && (string)value == "fruit", "SearchWithValue string");
                Assert(!trie.SearchWithValue("missing", out value), "SearchWithValue not found");

                // Generic version
                int intVal;
                Assert(trie.SearchWithValue<int>("apple", out intVal) && intVal == 1, "SearchWithValue<T> int");
                string strVal;
                Assert(trie.SearchWithValue<string>("banana", out strVal) && strVal == "fruit", "SearchWithValue<T> string");

                // GetCount
                trie.Insert("freq");
                trie.Insert("freq");
                trie.Insert("freq");
                Assert(trie.GetCount("freq") == 3, "GetCount returns frequency");
                Assert(trie.GetCount("missing") == 0, "GetCount for missing word");
            }
        }

        static void TestPrefixOperations()
        {
            Console.WriteLine("\nPrefix Operations Tests:");

            using (var trie = new Trie())
            {
                trie.Insert("hello");
                trie.Insert("help");
                trie.Insert("helper");
                trie.Insert("helicopter");
                trie.Insert("world");

                Assert(trie.StartsWith("hel"), "StartsWith existing prefix");
                Assert(trie.StartsWith("hello"), "StartsWith full word");
                Assert(trie.StartsWith(""), "StartsWith empty prefix");
                Assert(!trie.StartsWith("xyz"), "StartsWith non-existing prefix");

                var words = trie.WordsWithPrefix("hel");
                Assert(words.Count == 4, "WordsWithPrefix returns 4 words");
                Assert(words.Contains("hello") && words.Contains("help"), "WordsWithPrefix contains expected words");

                words = trie.WordsWithPrefix("xyz");
                Assert(words.Count == 0, "WordsWithPrefix non-existing returns empty");

                words = trie.WordsWithPrefixLimit("hel", 2);
                Assert(words.Count == 2, "WordsWithPrefixLimit respects limit");

                words = trie.AllWords();
                Assert(words.Count == 5, "AllWords returns all 5 words");
            }
        }

        static void TestDeleteOperations()
        {
            Console.WriteLine("\nDelete Operations Tests:");

            using (var trie = new Trie())
            {
                trie.Insert("hello");
                trie.Insert("help");
                trie.Insert("world");

                Assert(trie.Delete("hello"), "Delete existing word");
                Assert(trie.Size == 2, "Size decreased after delete");
                Assert(!trie.Search("hello"), "Deleted word not found");
                Assert(trie.Search("help"), "Other words still exist");

                Assert(!trie.Delete("hello"), "Delete non-existing word returns false");
                Assert(!trie.Delete(""), "Delete empty string returns false");
                Assert(!trie.Delete(null), "Delete null returns false");

                trie.Delete("help");
                trie.Delete("world");
                Assert(trie.Size == 0, "All deleted, size is 0");

                // Test cleanup of intermediate nodes
                trie.Insert("cat");
                trie.Insert("car");
                trie.Delete("cat");
                Assert(trie.Search("car"), "car still exists after cat deleted");
            }
        }

        static void TestAutoComplete()
        {
            Console.WriteLine("\nAutoComplete Tests:");

            using (var trie = new Trie())
            {
                trie.Insert("apple");
                trie.Insert("app");
                trie.Insert("application");
                trie.Insert("apply");
                trie.Insert("approach");
                trie.Insert("banana");

                var suggestions = trie.AutoComplete("app", 3);
                Assert(suggestions.Count <= 3, "AutoComplete respects limit");
                Assert(suggestions.All(s => s.StartsWith("app")), "All suggestions start with prefix");

                suggestions = trie.AutoComplete("app", 10);
                Assert(suggestions.Count == 5, "AutoComplete returns all matching words");

                suggestions = trie.AutoComplete("xyz", 5);
                Assert(suggestions.Count == 0, "AutoComplete non-existing prefix");

                // Batch insert
                var batchWords = new List<string> { "dog", "door", "dorm" };
                var count = trie.BatchInsert(batchWords);
                Assert(count == 3, "BatchInsert returns count of new words");

                suggestions = trie.AutoComplete("do", 5);
                Assert(suggestions.Count == 3, "AutoComplete finds batch inserted words");
            }
        }

        static void TestAdvancedFeatures()
        {
            Console.WriteLine("\nAdvanced Features Tests:");

            using (var trie = new Trie())
            {
                // LongestCommonPrefix
                trie.Insert("flower");
                trie.Insert("flow");
                trie.Insert("flight");
                Assert(trie.LongestCommonPrefix() == "fl", "LongestCommonPrefix");

                trie.Clear();
                trie.Insert("cat");
                trie.Insert("dog");
                Assert(trie.LongestCommonPrefix() == "", "LongestCommonPrefix no common");

                trie.Clear();
                trie.Insert("single");
                Assert(trie.LongestCommonPrefix() == "single", "LongestCommonPrefix single word");

                // MinPrefix
                trie.Clear();
                trie.Insert("cat");
                trie.Insert("car");
                trie.Insert("dog");
                Assert(trie.MinPrefix("cat") == "c", "MinPrefix for cat");
                Assert(trie.MinPrefix("dog") == "d", "MinPrefix for dog");

                // ContainsAnyPrefixOf
                trie.Clear();
                trie.Insert("car");
                trie.Insert("hello");
                Assert(trie.ContainsAnyPrefixOf("carpet"), "ContainsAnyPrefixOf finds car in carpet");
                Assert(!trie.ContainsAnyPrefixOf("xyz"), "ContainsAnyPrefixOf no match");

                // LongestPrefixOf
                trie.Insert("carpet");
                Assert(trie.LongestPrefixOf("carpets") == "carpet", "LongestPrefixOf");

                // PatternMatch
                trie.Clear();
                trie.Insert("cat");
                trie.Insert("bat");
                trie.Insert("rat");
                trie.Insert("car");
                trie.Insert("bar");

                var matches = trie.PatternMatch("?at");
                Assert(matches.Count == 3, "PatternMatch ?at finds 3 words");
                Assert(matches.Contains("cat") && matches.Contains("bat"), "PatternMatch contains expected");

                matches = trie.PatternMatch("ca?");
                Assert(matches.Count == 2, "PatternMatch ca? finds 2 words");

                matches = trie.PatternMatch("ca*");
                Assert(matches.Count >= 2, "PatternMatch ca* finds matching words");

                // GetWordsByFrequency
                trie.Clear();
                trie.Insert("apple");
                trie.Insert("apple");
                trie.Insert("apple");
                trie.Insert("banana");
                trie.Insert("banana");
                trie.Insert("cherry");

                var freq = trie.GetWordsByFrequency();
                Assert(freq.Count == 3, "GetWordsByFrequency returns 3 words");
                Assert(freq[0].Word == "apple" && freq[0].Count == 3, "GetWordsByFrequency sorted by freq");

                // ToMap
                trie.Clear();
                trie.Insert("a", 1);
                trie.Insert("b", 2);
                var map = trie.ToMap();
                Assert(map.Count == 2 && (int)map["a"] == 1, "ToMap works");

                // BatchInsertWithValues
                trie.Clear();
                var pairs = new Dictionary<string, object> { { "x", 10 }, { "y", 20 } };
                count = trie.BatchInsertWithValues(pairs);
                Assert(count == 2, "BatchInsertWithValues returns count");
                Assert(trie.SearchWithValue("x", out value) && (int)value == 10, "BatchInsertWithValues stored values");
            }
        }

        static void TestGenericTrie()
        {
            Console.WriteLine("\nGeneric Trie Tests:");

            using (var trie = new Trie<int>())
            {
                Assert(trie.Size == 0, "Generic trie empty");

                trie.Insert("score1", 100);
                trie.Insert("score2", 200);

                Assert(trie.Size == 2, "Generic trie size");

                int value;
                Assert(trie.Search("score1", out value) && value == 100, "Generic Search with value");
                Assert(trie.Search("score1"), "Generic Search without value");
                Assert(trie.StartsWith("score"), "Generic StartsWith");

                Assert(trie.Delete("score1"), "Generic Delete");
                Assert(!trie.Search("score1"), "Generic Delete verified");

                var words = trie.AutoComplete("sc", 5);
                Assert(words.Count == 1, "Generic AutoComplete");

                trie.Clear();
                Assert(trie.Size == 0, "Generic Clear");
            }
        }

        static void TestUnicodeSupport()
        {
            Console.WriteLine("\nUnicode Support Tests:");

            using (var trie = new Trie())
            {
                // Chinese
                trie.Insert("你好");
                trie.Insert("你好吗");
                trie.Insert("你们好");

                Assert(trie.Search("你好"), "Chinese word search");
                var words = trie.WordsWithPrefix("你好");
                Assert(words.Count == 2, "Chinese prefix search");

                // Japanese
                trie.Insert("こんにちは");
                Assert(trie.Search("こんにちは"), "Japanese word search");

                // Emoji
                trie.Insert("😀🎉");
                Assert(trie.Search("😀🎉"), "Emoji search");

                trie.Clear();
                Assert(trie.Size == 0, "Clear after Unicode test");
            }
        }

        static void TestConcurrency()
        {
            Console.WriteLine("\nConcurrency Tests:");

            using (var trie = new Trie())
            {
                var tasks = new List<System.Threading.Tasks.Task>();

                // Concurrent inserts
                for (int i = 0; i < 100; i++)
                {
                    tasks.Add(System.Threading.Tasks.Task.Run(() =>
                    {
                        trie.Insert("word" + Guid.NewGuid().ToString().Substring(0, 8));
                    }));
                }

                // Concurrent searches
                for (int i = 0; i < 50; i++)
                {
                    tasks.Add(System.Threading.Tasks.Task.Run(() =>
                    {
                        trie.Search("test");
                        trie.StartsWith("word");
                    }));
                }

                System.Threading.Tasks.Task.WaitAll(tasks.ToArray());
                Assert(trie.Size >= 100, "Concurrent inserts successful");

                // Concurrent reads and writes
                var finalSize = trie.Size;
                Assert(finalSize >= 100, "Concurrent operations completed");
            }
        }

        static void TestEdgeCases()
        {
            Console.WriteLine("\nEdge Cases Tests:");

            using (var trie = new Trie())
            {
                // Single character
                trie.Insert("a");
                Assert(trie.Search("a"), "Single character search");

                // Very long word
                var longWord = "abcdefghijklmnopqrstuvwxyz";
                trie.Insert(longWord);
                Assert(trie.Search(longWord), "Long word search");

                // Words with same prefix
                trie.Clear();
                trie.Insert("a");
                trie.Insert("aa");
                trie.Insert("aaa");
                Assert(trie.Search("a") && trie.Search("aa") && trie.Search("aaa"), "Nested prefixes");

                // Delete leaf
                trie.Delete("aaa");
                Assert(!trie.Search("aaa") && trie.Search("aa"), "Delete leaf preserves parent");

                // Empty trie operations
                trie.Clear();
                Assert(trie.AllWords().Count == 0, "AllWords on empty trie");
                Assert(trie.WordsWithPrefix("x").Count == 0, "WordsWithPrefix on empty trie");
                Assert(trie.LongestCommonPrefix() == "", "LongestCommonPrefix on empty trie");

                // Pattern match edge cases
                trie.Insert("a");
                trie.Insert("b");
                trie.Insert("c");
                var matches = trie.PatternMatch("*");
                Assert(matches.Count == 3, "PatternMatch * matches all");

                trie.Clear();
                trie.Insert("cat");
                trie.Insert("car");
                trie.Insert("card");
                trie.Insert("care");
                trie.Insert("careful");

                Assert(trie.LongestPrefixOf("carefully") == "careful", "LongestPrefixOf complex");
                Assert(trie.ContainsAnyPrefixOf("cardboard"), "ContainsAnyPrefixOf complex");
            }
        }
    }
}